### PR TITLE
feat(libs): add `tlv` and `tlv-derive` libs

### PR DESCRIPTION
**Pull request by eventualbuddha**
## Overview

These libraries provide TLV encode/decode functionality.

## Demo Video or Screenshot

An example of `derive`ing `tlv::Decode` and `tlv::Encode` and basic usage:
```rust
#[derive(tlv::Decode, tlv::Encode)]
struct Test {
    #[tlv(tag = 0x20)]
    a: u8,
    #[tlv(tag = 0x21)]
    b: u16,
}

let value = Test { a: 0, b: 999 };
let encoded = tlv::to_vec(&value)?;
let decoded: Test = tlv::from_slice(&encoded)?;
```

This is what the derive macros actually expand to (courtesy of `cargo expand`):

```rust
use tlv::Encode;
use tlv_derive::{Decode, Encode};
struct Test {
    #[tlv(tag = 0x20)]
    a: u8,
    #[tlv(tag = 0x21)]
    b: u16,
}
impl tlv::Encode for Test {
    fn encode<W>(&self, writer: &mut W) -> std::io::Result<()>
    where
        W: std::io::Write,
    {
        tlv::value::encode_tagged(tlv::Tag::U8(32u8), &self.a, writer)?;
        tlv::value::encode_tagged(tlv::Tag::U8(33u8), &self.b, writer)?;
        Ok(())
    }
    fn encoded_length(&self) -> std::io::Result<tlv::Length> {
        let mut length = tlv::Length::new(0);
        length += tlv::value::length_tagged(tlv::Tag::U8(32u8), &self.a)?;
        length += tlv::value::length_tagged(tlv::Tag::U8(33u8), &self.b)?;
        Ok(length)
    }
}
impl tlv::Decode for Test {
    fn decode<R>(reader: &mut R) -> std::io::Result<(Self, usize)>
    where
        R: std::io::Read,
    {
        let __tlv_decode_read: usize = 0;
        let (a, __tlv_decode_read) = match tlv::value::decode_tagged(
            tlv::Tag::U8(32u8),
            reader,
        )? {
            (value, read) => (value, read + __tlv_decode_read),
        };
        let (b, __tlv_decode_read) = match tlv::value::decode_tagged(
            tlv::Tag::U8(33u8),
            reader,
        )? {
            (value, read) => (value, read + __tlv_decode_read),
        };
        Ok((Self { a, b }, __tlv_decode_read))
    }
}
```

## Testing Plan
Adds a bunch of automated tests.

### CODE REVIEW

File: libs/tlv-derive/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "tlv-derive"
+version = "0.1.0"
+edition = "2021"
+
+[lib]
+proc-macro = true
```
Comment (eventualbuddha) [inline libs/tlv-derive/Cargo.toml:7]: I learned why `serde_derive` is its own crate and not part of `serde`. If your crate has procedural macros, then that's all it can be. You can't mix proc macro exports with non-proc macro exports. See https://doc.rust-lang.org/reference/procedural-macros.html.

File: libs/tlv-derive/src/lib.rs
Comment (eventualbuddha) [inline libs/tlv-derive/src/lib.rs:1]: This file is the one that actually provides the derive macros.

File: libs/tlv-derive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    };
+    let tags = entities
+        .iter()
+        .map(|e| match e.tag {
+            tlv::Tag::U8(tag) => quote! { tlv::Tag::U8(#tag) },
+            tlv::Tag::U16(tag) => quote! { tlv::Tag::U16(#tag) },
+        })
+        .collect::<Vec<_>>();
+    let idents = entities.iter().map(|e| &e.ident).collect::<Vec<_>>();
+    let gen = quote! {
```
Comment (eventualbuddha) [inline libs/tlv-derive/src/lib.rs:38]: The `quote!` macro allows building token streams with interpolation of syntax nodes.

File: libs/tlv-derive/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        impl tlv::Encode for #name {
```
Comment (eventualbuddha) [inline libs/tlv-derive/src/lib.rs:39]: This is a simple interpolation of a single value.

File: libs/tlv-derive/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            fn encode<W>(&self, writer: &mut W) -> std::io::Result<()>
+            where
+                W: std::io::Write,
+            {
+                #(tlv::value::encode_tagged(#tags, &self.#idents, writer)?;)*
```
Comment (eventualbuddha) [inline libs/tlv-derive/src/lib.rs:44]: This is a more complex interpolation using the `#(…)*` form that interpolates from both `#tags` and `#idents` at the same time. Think of it as zipping them together, using `tags[0]` with `idents[0]`, etc.

File: libs/tlv-derive/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        impl tlv::Decode for #name {
+            fn decode<R>(reader: &mut R) -> std::io::Result<(Self, usize)>
+            where
+                R: std::io::Read
+            {
+                let __tlv_decode_read: usize = 0;
+                #(
+                    let (#idents, __tlv_decode_read) = match tlv::value::decode_tagged(#tags, reader)? {
+                        (value, read) => (value, read + __tlv_decode_read),
+                    };
```
Comment (eventualbuddha) [inline libs/tlv-derive/src/lib.rs:93]: This takes advantage of the fact that Rust lets you redeclare a variable with the same name, shadowing the previous declaration, using the old one in the initializer.

File: libs/tlv-derive/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        tlv::Tag::U16(tag) => quote! { tlv::Tag::U16(#tag) },
+    });
+    let idents = entities.iter().map(|e| &e.ident).collect::<Vec<_>>();
+    let gen = quote! {
+        impl tlv::Decode for #name {
+            fn decode<R>(reader: &mut R) -> std::io::Result<(Self, usize)>
+            where
+                R: std::io::Read
+            {
+                let __tlv_decode_read: usize = 0;
```
Comment (eventualbuddha) [inline libs/tlv-derive/src/lib.rs:89]: It'd be nice to generate a known-safe identifier here, but I think this is probably sufficient. Nobody's going to use this name for a field. If I did want to, I think I'd have to just use a random number and validate that `idents` doesn't contain that value.

File: libs/tlv/src/coder.rs [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+use crate::length::Length;
+
+/// Encodes a value into a writer.
+pub trait Encode {
+    /// Converts the value into a byte representation and writes it to the
+    /// writer.
+    fn encode<W>(&self, writer: &mut W) -> std::io::Result<()>
+    where
+        W: Write;
```
Comment (eventualbuddha) [inline libs/tlv/src/coder.rs:11]: In an earlier version of this I had the generic parameter `W` at the trait level, but that made it harder to use in other places because any function that accepted an `Encode` also had to have the `W` as its own type parameter. This made, for example, implementing `tlv::to_vec` difficult or impossible.

File: libs/tlv/src/coder.rs [hunk 8]
```diff
@@ -0,0 +1,165 @@
+use std::io::{Read, Write};
+
+use crate::length::Length;
+
+/// Encodes a value into a writer.
+pub trait Encode {
+    /// Converts the value into a byte representation and writes it to the
+    /// writer.
+    fn encode<W>(&self, writer: &mut W) -> std::io::Result<()>
```
Comment (eventualbuddha) [inline libs/tlv/src/coder.rs:9]: I went with a fairly uninformative `std::io::Result` for these rather than a more specific `Error`, but if I were to improve this library that's likely something I'd do to include things like field names, types, offsets, etc. a la `serde`.

File: libs/tlv/src/coder.rs [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                    let mut buf = [0; SIZE];
+                    reader.read_exact(&mut buf)?;
+                    Ok((<$t>::from_be_bytes(buf), SIZE))
+                }
+            }
+        )*
+    };
+}
+
+impl_coders!(u8, u16, u32, u64, usize, i8, i16, i32, i64, isize);
```
Comment (eventualbuddha) [inline libs/tlv/src/coder.rs:53]: 💥

File: libs/tlv/src/coder.rs [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+impl<T: Encode> Encode for &T {
```
Comment (eventualbuddha) [inline libs/tlv/src/coder.rs:55]: This is a useful trick for implementing a trait for a whole class of thing. This says "for all `T` that implement `Encode`, here's how to implement `Encode` for `&T`". This means you can just accept `Encode` values and the caller can pass either an owned `T` or a borrowed `&T`:

```rust
let value = Test { a: 99 };
tlv::to_vec(&value); // just borrowing, I want to use it again
tlv::to_vec(value); // okay, I'm done with `value` so pass ownership
```

File: libs/tlv-derive/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+                #(
+                    let (#idents, __tlv_decode_read) = match tlv::value::decode_tagged(#tags, reader)? {
```
Comment (eventualbuddha) [inline libs/tlv-derive/src/lib.rs:91]: A nice thing about this is that I don't have to specify the type of `#idents` here, either on the `let` or as a type parameter to `decode_tagged`. Rust will work backward from the returned `Ok((Self { … }, …))` to determine the expected type of each field, making it much easier to derive an implementation because the types can be omitted.

File: libs/tlv/src/coder.rs [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        Vec::<u8>::decode(reader).and_then(|(data, read)| {
+            match std::string::String::from_utf8(data) {
+                Ok(string) => Ok((string, read)),
+                Err(_) => Err(std::io::ErrorKind::InvalidData.into()),
+            }
+        })
+    }
+}
+
+impl Decode for () {
```
Comment (eventualbuddha) [inline libs/tlv/src/coder.rs:158]: Useful edge case to test. It's probably a good idea to implement for more tuples than the empty tuple, but unfortunately Rust doesn't have a mechanism for defining an implementation for arbitrary-length tuples.

File: libs/tlv/src/lib.rs
Comment (eventualbuddha) [inline libs/tlv/src/lib.rs:1]: The free functions implemented in this module are loosely based on the equivalents from `serde_json`.

File: libs/tlv/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+
+        let (fe, read) = from_reader_tagged::<u8, _>(Tag::U8(0xff), &mut cursor).unwrap();
+        assert_eq!(fe, 0xfe);
+        assert_eq!(read, 3);
+
+        assert_eq!(cursor.position(), data.len() as u64);
+    }
+
+    #[test]
+    fn test_encode_codegen() {
```
Comment (eventualbuddha) [inline libs/tlv/src/lib.rs:209]: This and the decode test below aren't strictly necessary, but were useful when I was figuring out how to implement `tlv-derive`.

File: libs/tlv/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+        assert_eq!(decoded, value);
+    }
+
+    #[test]
+    fn test_boolean() {
+        assert_eq!(to_vec(true).unwrap(), [0x01]);
+        assert_eq!(to_vec(false).unwrap(), [0x00]);
+    }
+
+    proptest! {
```
Comment (eventualbuddha) [inline libs/tlv/src/lib.rs:299]: The test functions in this block will be called with automatic random values for each of the arguments, similar to `fast-check` in JS but with less boilerplate.

File: libs/tlv/src/coder.rs (hunk 8 above)
Comment (eventualbuddha) [reply libs/tlv/src/coder.rs]: This has been fixed.

File: libs/tlv/src/value.rs [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+    if tag_bytes != read_tag_bytes {
+        return Err(Error::InvalidTag {
+            expected: tag_bytes,
+            actual: read_tag_bytes,
+        });
+    }
+
+    let length = Length::from_reader(reader)?;
+    let (value, value_size) = D::decode(&mut LimitedReader::new(reader, length.value().into()))?;
```
Comment (eventualbuddha) [inline libs/tlv/src/value.rs:44]: Adding the `LimitedReader` wrapper fixes an issue where variable-length values would just read all data from the reader until the end, leaving nothing for any remaining decoders that were supposed to read after them. This limits what a decoder can read to the encoded length.
